File: ios/Classes/WannatalkcorePlugin.m
#import "WannatalkcorePlugin.h"
#import <WTExternalSDK/WTExternalSDK.h>
#import "WTConfigHandler.h"


@interface WannatalkcorePlugin() <WTLoginManagerDelegate, WTSDKManagerDelegate> {
    BOOL _hasListeners;



}
@property (nonatomic, strong) FlutterResult loginResult;
@property (nonatomic, strong) FlutterResult logoutResult;
@property (nonatomic, strong) FlutterResult orgProfileResult;
@property (nonatomic, strong) FlutterResult chatListResult;
@property (nonatomic, strong) FlutterResult userListResult;


@property (nonatomic, strong) FlutterMethodChannel *wtEventChannel;

@end

@implementation WannatalkcorePlugin


+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    WannatalkcorePlugin* instance = [[WannatalkcorePlugin alloc] init];
    instance.wtEventChannel = [FlutterMethodChannel methodChannelWithName:@"wannatalkcore" binaryMessenger:[registrar messenger]];
    [registrar addMethodCallDelegate:instance channel:instance.wtEventChannel];
    [instance initialize];

}

- (void) initialize {
    _hasListeners = YES;
    [[WTSDKApplicationDelegate sharedInstance] application:[UIApplication sharedApplication] didFinishLaunchingWithOptions:nil];

    [WTLoginManager sharedInstance].delegate = self;
    [WTSDKManager sharedInstance].delegate = self;
}


#pragma mark - Constants

static const  NSString * _Nonnull _cWTDefaultMethod= @"wannatalkDefaultMethod";
static const  NSString * _Nonnull _cWTUpdateConfig= @"updateConfig";
static const  NSString * _Nonnull _cWTIsUserLoggedIn= @"isUserLoggedIn";


static const  NSString * _Nonnull _cWTMethodType= @"methodType";
static const  NSString * _Nonnull _cWTArgs= @"args";
static const  NSString * _Nonnull _cWTUsername= @"username";
static const  NSString * _Nonnull _cWTLocalImagePath= @"localImagePath";

static const  NSString * _Nonnull _cWTAutoOpenChat= @"autoOpenChat";
static const  NSString * _Nonnull _cWTUserIdentifier= @"userIdentifier";
static const  NSString * _Nonnull _cWTUserInfo= @"userInfo";


static const  int _kWTLoginMethod= 1001;
static const  int _kWTSilentLoginMethod= 1002;
static const  int _kWTLogoutMethod= 1003;
static const  int _kWTOrgProfileMethod= 1004;
static const  int _kWTChatListMethod= 1005;
static const  int _kWTUserListMethod= 1006;
static const  int _kWTUpdateUserNameMethod= 1007;
static const  int _kWTUpdateUserImageMethod= 1008;
static const  int _kWTIsUserLoggedIn= 1009;


// OUT

static const  NSString * _Nonnull _cMethod= @"wt_event";
static const  NSString * _Nonnull _cEventType= @"eventType";
static const  NSString * _Nonnull _cSuccess= @"success";
static const  NSString * _Nonnull _cError= @"error";

static const  int _kWTEventTypeLogin= 2001;
static const  int _kWTEventTypeLogout= 2002;


#pragma mark -

- (void) handleMethodType:(NSInteger) methodType arguments:(NSDictionary *) args result:(FlutterResult)result {
    switch (methodType) {
        case _kWTLoginMethod: {
            [self login:result];
            break;
        }
        case _kWTSilentLoginMethod: {
            NSString *identifier = args[_cWTUserIdentifier];
            NSDictionary *userInfo = args[_cWTUserInfo];
            [self silentLogin:identifier userInfo:userInfo result:result];

            break;
        }
        case _kWTLogoutMethod: {
            [self logout:result];
            break;
        }
        case _kWTOrgProfileMethod: {
            BOOL autoOpenChat = [args[_cWTAutoOpenChat] boolValue];
            [self loadOrganizationProfile:autoOpenChat result:result];

            break;
        }
        case _kWTChatListMethod: {
            [self loadChatList:result];
            break;
        }
        case _kWTUserListMethod: {
            [self loadUsers:result];
            break;
        }
        case _kWTUpdateUserNameMethod: {
            NSString *username = args[_cWTUsername];
            [self updateUserName:username result:result];
            break;
        }
        case _kWTUpdateUserImageMethod: {
            NSString *localImagePath = args[_cWTLocalImagePath];
            [self updateUserImage:localImagePath result:result];

            break;
        }
        default: {

            break;
        }
    }
}

- (void) handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result {
    NSString *method = call.method;

    if ([method isEqualToString:_cWTDefaultMethod]) {

        NSInteger methodType = [call.arguments[_cWTMethodType] integerValue];
        NSDictionary *args = call.arguments[_cWTArgs];
        [self handleMethodType:methodType arguments:args result:result];

    }
    else if ([method isEqualToString:_cWTUpdateConfig]) {
        NSInteger methodType = [call.arguments[_cWTMethodType] integerValue];
        NSDictionary *args = call.arguments[_cWTArgs];
        [WTConfigHandler handleConfigMethodType:methodType arguments:args];
        result(@(YES));
    }
    else if ([method isEqualToString:_cWTIsUserLoggedIn]) {
        BOOL loggedIn = [self isUserLoggedIn];
        result(@(loggedIn));
    }
}



#pragma mark -

+ (UIViewController *) GetWindowRootViewController {
    UIWindow *window = [[UIApplication sharedApplication] keyWindow];
    UIViewController *_rootViewController = (UIViewController *)window.rootViewController;
    return _rootViewController;
}

+ (void) RunMainThread:(void (^)(void))onMainThread {

    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        dispatch_async(dispatch_get_main_queue(), ^{
            if (onMainThread) {
                onMainThread();
            }
        });
    });

}

+ (void) GetBaseViewController:(void (^)(UIViewController *baseViewController)) onCompletion {

    [self RunMainThread:^{
        UIViewController *_rootViewController = [self GetWindowRootViewController];

        if (onCompletion) {
            onCompletion(_rootViewController);
        }
    }];


}
#pragma mark -

- (BOOL) isUserLoggedIn {
    return [WTLoginManager sharedInstance].isUserLoggedIn;
}

- (void) logout:(FlutterResult) result {
    self.logoutResult = result;

    [WannatalkcorePlugin RunMainThread:^{
        [[WTLoginManager sharedInstance] logout];
    }];
}

- (void)silentLogin:(NSString *) identifier userInfo:(NSDictionary *) userInfo result:(FlutterResult) result {
    self.loginResult = result;

    [WannatalkcorePlugin GetBaseViewController:^(UIViewController *baseViewController) {
        [[WTLoginManager sharedInstance] silentLoginWithIdentifier:identifier userInfo:userInfo fromVC:baseViewController];

    }];
}


-(void) login:(FlutterResult) result {

    self.loginResult = result;

    [WannatalkcorePlugin GetBaseViewController:^(UIViewController *baseViewController) {
        [[WTLoginManager sharedInstance] loginFromVC:baseViewController];

    }];

}

#pragma mark -

- (void)loadOrganizationProfile:(BOOL) autoOpenChat result:(FlutterResult) result {
    self.orgProfileResult = result;

    [WannatalkcorePlugin GetBaseViewController:^(UIViewController *baseViewController) {
        [baseViewController presentOrgProfileVCWithAutoOpenChat:autoOpenChat delegate:self animated:YES completion:nil];
    }];

}

-(void) loadChatList:(FlutterResult) result {
    self.chatListResult = result;
    [WannatalkcorePlugin GetBaseViewController:^(UIViewController *baseViewController) {
        [baseViewController presentChatListVCWithDelegate:self animated:YES completion:nil];
    }];
}


- (void)loadUsers:(FlutterResult) result {
    self.userListResult = result;

    [WannatalkcorePlugin GetBaseViewController:^(UIViewController *baseViewController) {
        [baseViewController presentUsersVCWithDelegate:self animated:YES completion:nil];
    }];
}

#pragma mark -

- (void) updateUserImage:(NSString *) localImagePath result:(FlutterResult) result {
    [[WTLoginManager sharedInstance] uploadUserImageAtPath:localImagePath onCompletion:^(BOOL success, NSString *error) {

        if (result) {
            result(error);
        }

    }];
}

- (void) updateUserName:(NSString *) userName result:(FlutterResult) result {
    [[WTLoginManager sharedInstance] updateUserProfileName:userName onCompletion:^(BOOL success, NSString *error) {

        if (result) {
            result(error);
        }
    }];
}


#pragma mark -


- (void) sendWannatalkEvent:(NSInteger ) eventType error:(NSString *) error {
    if (_hasListeners) {
        NSMutableDictionary *args = [NSMutableDictionary new];
        args[_cEventType] = @(eventType);
        args[_cSuccess] = @(error == nil);
        args[_cError] = error;

        [self.wtEventChannel invokeMethod:_cMethod arguments:args];
    }
}

- (void) sendLoginCallback:(NSString *) error {
    if (self.loginResult) {
        self.loginResult(error);
    }
    self.loginResult = nil;
}

- (void) sendLogoutCallback:(NSString *) error {
    if (self.logoutResult) {
        self.logoutResult(error);
    }
    self.logoutResult = nil;
}


- (void) sendOrgProfileCallback:(NSString *) error {
    if (self.orgProfileResult) {
        self.orgProfileResult(error);
    }
    self.orgProfileResult = nil;
}


- (void) sendChatListCallback:(NSString *) error {
    if (self.chatListResult) {
        self.chatListResult(error);
    }
    self.chatListResult = nil;
}


- (void) sendUserListallback:(NSString *) error {
    if (self.userListResult) {
        self.userListResult(error);
    }
    self.userListResult = nil;
}


#pragma mark - WTLoginManager Delegate Methods

// This method will be invoked when user sign in successfully
- (void) wtAccountDidLoginSuccessfully {
    [self sendLoginCallback:nil];
    [self sendWannatalkEvent:_kWTEventTypeLogin error:nil];
}
// This method will be invoked when user sign out successfully
- (void) wtAccountDidLogOutSuccessfully {
    [self sendLogoutCallback:nil];
    [self sendWannatalkEvent:_kWTEventTypeLogout error:nil];
}

// If implemented, this method will be invoked when login fails
- (void) wtAccountDidLoginFailWithError:(NSString *) error {
    [self sendLoginCallback:error];
    [self sendWannatalkEvent:_kWTEventTypeLogin error:error];
}
// If implemented, this method will be invoked when logout fails
- (void) wtAccountDidLogOutFailedWithError:(NSString *) error {
    [self sendLogoutCallback:error];
    [self sendWannatalkEvent:_kWTEventTypeLogout error:error];
}

#pragma mark - WTSDKManager Delegate Methods

- (UINavigationController *) prepareViewHierachiesToLoadChatRoom:(BOOL) aiTopic {
    // Get view controller on which to present the flow
    UINavigationController *_rootViewController = (UINavigationController *) [WannatalkcorePlugin GetWindowRootViewController];
    return _rootViewController;
}


// If implemented, this method will be invoked when organization profile loads successfully
- (void) wtsdkOrgProfileDidLoadSuccesfully {
    [self sendOrgProfileCallback:nil];
}

// If implemented, this method will be invoked when organization profile fails to load
- (void) wtsdkOrgProfileDidLoadFailWithError:(NSString *) error {
    [self sendOrgProfileCallback:error];
}

// If implemented, this method will be invoked when chat list page loads successfully
- (void) wtsdkChatListDidLoadSuccesfully {
    [self sendChatListCallback:nil];

}

// If implemented, this method will be invoked when chat list page fails to load
- (void) wtsdkChatListDidLoadFailWithError:(NSString *) error {
    [self sendChatListCallback:error];
}

// If implemented, this method will be invoked when user list page loads successfully
- (void) wtsdkUsersDidLoadSuccesfully {
    [self sendUserListallback:nil];
}

// If implemented, this method will be invoked when user list page fails to load
- (void) wtsdkUsersDidLoadFailWithError:(NSString *) error {
    [self sendUserListallback:error];
}
@end

File: android/src/main/java/com/wannatalk/wannatalkcore/WannatalkcorePlugin.java
package com.wannatalk.wannatalkcore;


import android.support.annotation.NonNull;
import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.embedding.engine.plugins.activity.ActivityAware;

import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

import wannatalk.wannatalksdk.WTCore.Interface.IWTLoginManager;
import wannatalk.wannatalksdk.WTCore.Interface.IWTCompletion;
import wannatalk.wannatalksdk.WTCore.WTSDKManager;
import wannatalk.wannatalksdk.WTCore.WTSDKConstants;
import wannatalk.wannatalksdk.WTLogin.WTLoginManager;

import android.util.Log;

import android.app.Activity;
import android.os.Bundle;

import java.util.HashMap;
import java.util.Map;

/** WannatalkcorePlugin */
public class WannatalkcorePlugin implements FlutterPlugin, ActivityAware, MethodCallHandler {
//  @Override
//  public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
//    final MethodChannel channel = new MethodChannel(flutterPluginBinding.getFlutterEngine().getDartExecutor(), "wannatalkcore");
//    channel.setMethodCallHandler(new WannatalkcorePlugin());
//  }
//
//  // This static function is optional and equivalent to onAttachedToEngine. It supports the old
//  // pre-Flutter-1.12 Android projects. You are encouraged to continue supporting
//  // plugin registration via this function while apps migrate to use the new Android APIs
//  // post-flutter-1.12 via https://flutter.dev/go/android-project-migration.
//  //
//  // It is encouraged to share logic between onAttachedToEngine and registerWith to keep
//  // them functionally equivalent. Only one of onAttachedToEngine or registerWith will be called
//  // depending on the user's project. onAttachedToEngine or registerWith must both be defined
//  // in the same class.
//  public static void registerWith(Registrar registrar) {
//    final MethodChannel channel = new MethodChannel(registrar.messenger(), "wannatalkcore");
//    channel.setMethodCallHandler(new WannatalkcorePlugin());
//  }
//
//  @Override
//  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
//    if (call.method.equals("getPlatformVersion")) {
//      result.success("Android " + android.os.Build.VERSION.RELEASE);
//    } else {
//      result.notImplemented();
//    }
//  }
//
//  @Override
//  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
//  }


  private static WannatalkcorePlugin plugin;
  private static MethodChannel channel;
  static Activity sActivity;
  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {

    if (channel == null || plugin == null) {
      channel = new MethodChannel(flutterPluginBinding.getFlutterEngine().getDartExecutor(), "wannatalkcore");
      plugin = new WannatalkcorePlugin();
      channel.setMethodCallHandler(plugin);
      plugin.initialize();
    }

  }

  // This static function is optional and equivalent to onAttachedToEngine. It supports the old
  // pre-Flutter-1.12 Android projects. You are encouraged to continue supporting
  // plugin registration via this function while apps migrate to use the new Android APIs
  // post-flutter-1.12 via https://flutter.dev/go/android-project-migration.
  //
  // It is encouraged to share logic between onAttachedToEngine and registerWith to keep
  // them functionally equivalent. Only one of onAttachedToEngine or registerWith will be called
  // depending on the user's project. onAttachedToEngine or registerWith must both be defined
  // in the same class.
  private static void registerWith(Registrar registrar) {
    if (channel == null || plugin == null) {
      channel = new MethodChannel(registrar.messenger(), "wannatalkcore");
      plugin = new WannatalkcorePlugin();
      channel.setMethodCallHandler(plugin);
      plugin.initialize();
    }
  }
  void initialize() {
    // Context context = getApplicationContext();
//    final Application applicationContext = (Application) this.reactContext.getApplicationContext();


    WTLoginManager.setIwtLoginManager(iwtLoginManager);
  }

  // IN
  static final  String _cWTDefaultMethod= "wannatalkDefaultMethod";
  static final  String _cWTUpdateConfig= "updateConfig";
  static final  String _cWTIsUserLoggedIn= "isUserLoggedIn";


  static final  String _cWTMethodType= "methodType";
  static final  String _cWTArgs= "args";
  static final  String _cWTUsername= "username";
  static final  String _cWTLocalImagePath= "localImagePath";

  static final  String _cWTAutoOpenChat= "autoOpenChat";
  static final  String _cWTUserIdentifier= "userIdentifier";
  static final  String _cWTUserInfo= "userInfo";


  static final  int _kWTLoginMethod= 1001;
  static final  int _kWTSilentLoginMethod= 1002;
  static final  int _kWTLogoutMethod= 1003;
  static final  int _kWTOrgProfileMethod= 1004;
  static final  int _kWTChatListMethod= 1005;
  static final  int _kWTUserListMethod= 1006;
  static final  int _kWTUpdateUserNameMethod= 1007;
  static final  int _kWTUpdateUserImageMethod= 1008;
  static final  int _kWTIsUserLoggedIn= 1009;


  // OUT

  static final  String _cMethod= "wt_event";
  static final  String _cEventType= "eventType";
  static final  String _cSuccess= "success";
  static final  String _cError= "error";

  static final  int kWTEventTypeLogin= 2001;
  static final  int kWTEventTypeLogout= 2002;


//  Activity activity;

  @Override
  public void onAttachedToActivity(ActivityPluginBinding binding) {
//    activity = binding.getActivity();
    sActivity = binding.getActivity();

  }

  @Override
  public void onDetachedFromActivityForConfigChanges() {

  }

  @Override
  public void onReattachedToActivityForConfigChanges(ActivityPluginBinding activityPluginBinding) {

//    activity = activityPluginBinding.getActivity();
    sActivity = activityPluginBinding.getActivity();
  }

  @Override
  public void onDetachedFromActivity() {
//    activity = null;
    sActivity = null;
  }

  public void onHandleMethodType(int methodType, Map<String, Object> args, @NonNull Result result) {
    switch (methodType) {
      case _kWTLoginMethod: {
        login(result);
        break;
      }
      case _kWTSilentLoginMethod: {
        String identifier = (String) args.get(_cWTUserIdentifier);
        Map<String, String> userInfo = (Map<String, String>) args.get(_cWTUserInfo);
        silentLogin(identifier, userInfo, result);
        break;
      }
      case _kWTLogoutMethod: {
        logout(result);
        break;
      }
      case _kWTOrgProfileMethod: {
        boolean autoOpenChat = (boolean) args.get(_cWTAutoOpenChat);
        loadOrganizationProfile(autoOpenChat, result);
        break;
      }
      case _kWTChatListMethod: {
        loadChatList(result);
        break;
      }
      case _kWTUserListMethod: {
        loadUsers(result);
        break;
      }
      case _kWTUpdateUserNameMethod: {
        String username = (String) args.get(_cWTUsername);
        updateUserName(username, result);
        break;
      }
      case _kWTUpdateUserImageMethod: {
        String localImagePath = (String) args.get(_cWTLocalImagePath);
        updateUserImage(localImagePath, result);
        break;
      }
      default: {
        result.notImplemented();
        break;
      }
    }
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
    if (call.method.equals(_cWTDefaultMethod)) {
      Map<String, Object> mainArgs = (Map<String, Object>) call.arguments;

      int methodType = (int) mainArgs.get(_cWTMethodType);
      Map<String, Object> args = (Map<String, Object>) mainArgs.get(_cWTArgs);
      onHandleMethodType(methodType, args, result);
    }
    else if (call.method.equals(_cWTUpdateConfig)) {
      Map<String, Object> mainArgs = (Map<String, Object>) call.arguments;

      int methodType = (int) mainArgs.get(_cWTMethodType);
      Map<String, Object> args = (Map<String, Object>) mainArgs.get(_cWTArgs);
      WTConfigHandler.HandleMethodType(methodType, args, result);
    }
    else if (call.method.equals(_cWTIsUserLoggedIn)) {

      Boolean loggedIn = isUserLoggedIn();
      result.success(loggedIn );
    }
    else {
      result.notImplemented();
    }
  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
  }



  public Boolean isUserLoggedIn() {
    return WTLoginManager.IsUserLoggedIn();
  }

  Activity getActivity() {

    return sActivity;
  }

  Result loginResult;
  Result logoutResult;
  void login(Result result) {
    loginResult = result;

    Activity currentActivity = getActivity();
    if (currentActivity != null) {
      WTLoginManager.StartLoginActivity(currentActivity);
    }
    else {
      sendLoginCallback("Unable to get the context");
    }
  }


  void silentLogin(final String identifier, final Map<String, String> userInfo, Result result) {
    loginResult = result;
    // Silent authentication without otp verification

    Activity currentActivity = getActivity();
    if (currentActivity != null) {
      Bundle bundle = new Bundle();
      if (userInfo != null) {
        for (Map.Entry<String,String> entry : userInfo.entrySet()) {
          String key = entry.getKey();
          String value = entry.getValue();
          bundle.putString(key, value);
        }
      }

      WTLoginManager.SilentLoginActivity(identifier, bundle, currentActivity);
    }
    else {
      sendLoginCallback("Unable to get the context");
    }



  }


  void logout(Result result) {
    logoutResult = result;
    Activity currentActivity = getActivity();
    if (currentActivity != null) {
      WTLoginManager.Logout(currentActivity);
    }
    else {
      sendLogoutCallback("Unable to get the context");
    }
  }

  void loadOrganizationProfile(Boolean autoOpenChat,final Result result) {
    // Load organization profile
    Activity currentActivity = getActivity();
    if (currentActivity != null) {
      WTSDKManager.LoadOrganizationActivity(currentActivity, autoOpenChat, new IWTCompletion() {
        @Override
        public void onCompletion(boolean success, String error) {
          if (success) {
            if (result != null) {
              result.success(null);
            }
          }
          else {
            if (result != null) {
              result.success(error);
            }
          }

        }
      });
    }
    else {
      if (result != null) {
        result.success("Unable to get the context");
      }

    }

  }


  void loadChatList(final Result result) {
    // Load chat list
    Activity currentActivity = getActivity();
    if (currentActivity != null) {
      WTSDKManager.LoadChatListActivity(currentActivity, new IWTCompletion() {
        @Override
        public void onCompletion(boolean success, String error) {
          if (success) {
            if (result != null) {
              result.success(null);
            }
          }
          else {
            if (result != null) {
              result.success(error);
            }
          }
        }
      });

    }
    else {
      if (result != null) {
        result.success("Unable to get the context");
      }

    }
  }


  void loadUsers(final Result result) {
    // Load users
    Activity currentActivity = getActivity();
    if (currentActivity != null) {
      WTSDKManager.LoadUsersActivity(currentActivity, new IWTCompletion() {
        @Override
        public void onCompletion(boolean success, String error) {
          if (success) {
            if (result != null) {
              result.success(null);
            }
          }
          else {
            if (result != null) {
              result.success(error);
            }
          }
        }
      });

    }
    else {
      if (result != null) {
        result.success("Unable to get the context");
      }

    }
  }

//  //
//  // private static void sendNotification(RemoteMessage remoteMessage, Context
//  // context) {
//  // WTBaseSDKManager.SendNotification(remoteMessage, context);
//  // }
//
//  //
//  // private static void SetDeviceToken(String deviceToken) {
//  // WTBaseSDKManager.SetDeviceToken(deviceToken);
//  // }

  private static void updateUserImage(String localImagePath, final Result result) {
    WTLoginManager.UploadUserImageAtPath(localImagePath, new IWTCompletion() {
      @Override
      public void onCompletion(boolean success, String error) {
        if (success) {
          if (result != null) {
            result.success(null);
          }
        }
        else {
          if (result != null) {
            result.success(error);
          }
        }
      }
    });
  }



  private static void updateUserName(String userName, final Result result) {
    WTLoginManager.UpdateUserProfileName(userName, new IWTCompletion() {
      @Override
      public void onCompletion(boolean success, String error) {
        if (success) {
          if (result != null) {
            result.success(null);
          }
        }
        else {
          if (result != null) {
            result.success(error);
          }
        }
      }
    });

  }


  private void sendLoginCallback(String error) {
    if (loginResult != null) {
      loginResult.success(error);
    }
    loginResult = null;
  }

  private void sendLogoutCallback(String error) {
    if (logoutResult != null) {
      logoutResult.success(error);
    }
    logoutResult = null;
  }

  private void sendWannatalkEvent(int eventType, String error) {

    Map<String, Object> args = new HashMap<>();
    args.put(_cEventType, eventType);
    args.put(_cSuccess,error == null);
    args.put(_cError, error);


    channel.invokeMethod(_cMethod, args);
  }


  IWTLoginManager iwtLoginManager = new IWTLoginManager() {

    @Override
    public void wtsdkUserLoggedIn() {
      sendLoginCallback(null);
      sendWannatalkEvent(kWTEventTypeLogin, null);
    }

    @Override
    public void wtsdkUserLoginFailed(String error) {
      sendLoginCallback(error);
      sendWannatalkEvent(kWTEventTypeLogin, error);
    }

    @Override
    public void wtsdkUserLoggedOut() {
      sendLogoutCallback(null);
      sendWannatalkEvent(kWTEventTypeLogout, null);
    }

    @Override
    public void wtsdkUserLogoutFailed(String error) {
      sendLogoutCallback(error);
      sendWannatalkEvent(kWTEventTypeLogout, error);
    }

  };
}
